#
# Submitted-By: Marc Kleine-Budde, 2007-04-04
# Committed-By: Marc Kleine-Budde
#
# Error:
#
# Corrent Thumb interworking support is broken in gcc-4.1.x
#
# State:
#
# This patch is from
# http://www.openembedded.org/repo/org.openembedded.dev/packages/gcc/gcc-4.1.2/arm-thumb-cache.patch
# Wed Apr  4 16:49:30 CEST 2007
#
---
# gcc/config/arm/linux-gas.h |   16 ++++++++++++++++
# 1 file changed, 16 insertions(+)
#
Index: gcc-4.2.1/gcc/config/arm/linux-gas.h
===================================================================
--- gcc-4.2.1.orig/gcc/config/arm/linux-gas.h	2007-04-16 22:16:13.000000000 +0200
+++ gcc-4.2.1/gcc/config/arm/linux-gas.h	2007-08-27 17:09:11.000000000 +0200
@@ -44,6 +44,7 @@
 
 /* Clear the instruction cache from `beg' to `end'.  This makes an
    inline system call to SYS_cacheflush.  */
+#if !defined(__thumb__)
 #define CLEAR_INSN_CACHE(BEG, END)					\
 {									\
   register unsigned long _beg __asm ("a1") = (unsigned long) (BEG);	\
@@ -53,3 +54,18 @@
 		    : "=r" (_beg)					\
 		    : "0" (_beg), "r" (_end), "r" (_flg));		\
 }
+#else
+#define CLEAR_INSN_CACHE(BEG, END)					\
+{									\
+  register unsigned long _beg __asm ("a1") = (unsigned long) (BEG);	\
+  register unsigned long _end __asm ("a2") = (unsigned long) (END);	\
+  register unsigned long _flg __asm ("a3") = 0;				\
+  register unsigned long _swi __asm ("a4") = 0xf0002;			\
+  __asm __volatile ("push      {r7}\n"                                 \
+               "       mov     r7,a4\n"                                \
+               "       swi     0       @ sys_cacheflush\n"             \
+               "       pop     {r7}\n"                                 \
+                   : "=r" (_beg)                                       \
+                   : "0" (_beg), "r" (_end), "r" (_flg), "r" (_swi));  \
+}
+#endif
